Regression(255052@main) Unable to log into Marcus.com on iOS

cdumez opened the pull request:
#### 3622b4c14b28cbd60481fe9952e40e4a4f83cfa7
<pre>
Regression(255052@main) Unable to log into Marcus.com on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=248798">https://bugs.webkit.org/show_bug.cgi?id=248798</a>
rdar://102959860

Reviewed by Brent Fulgham.

Since 255052@main, login on marcus.com fails on iOS and complains about being
outside the USA. The same thing happened by we dropped `window.showModalDialog()`.
The site&apos;s login seems to rely on which properties are present on the `window`
and `navigator` object and what their values are.

In this case, the log in fails because `navigator.plugins` and
`navigator.mimeTypes` now returns the values hard-coded in the HTML
specification instead of being empty arrays.

This is a site issue so I am adding a quirk for now and contacted evangelism.

* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::initializePluginAndMimeTypeArrays):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldNavigatorPluginsBeEmpty const):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/257407@main">https://commits.webkit.org/257407@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d50bcfe565eb4412e3874e60c989927760f7a4d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8083 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32029 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108287 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168544 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102827 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85449 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91406 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106272 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104555 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6542 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90115 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33572 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88364 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76428 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1993 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22986 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1902 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6858 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42449 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2576 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3301 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1474,6 +1474,22 @@ bool Quirks::shouldExposeShowModalDialog() const
+    if (!m_shouldNavigatorPluginsBeEmpty) {
+        auto domain = RegistrableDomain(m_document->url()).string();
+        // Marcus login issue: <rdar://102959860>.
+        m_shouldNavigatorPluginsBeEmpty = domain == "marcus.com"_s;
```
Let's add a console message for it Ala:

```cpp
m_document->addConsoleMessage(MessageSource::Other, MessageLevel::Info, "QUIRK: Navigator plugins empty on marcus.com. More information at https://bugs.webkit.org/show_bug.cgi?id=248798"_s);
```

or something probably cleverer than this code.

brentfulgham reviewed (approved):
Looks good to me. We should reference the Evangelism radar so we know when we can remove this code.

cdumez committed:
Regression(255052@main) Unable to log into Marcus.com on iOS
https://bugs.webkit.org/show_bug.cgi?id=248798
rdar://102959860

Reviewed by Brent Fulgham.

Since 255052@main, login on marcus.com fails on iOS and complains about being
outside the USA. The same thing happened by we dropped `window.showModalDialog()`.
The site's login seems to rely on which properties are present on the `window`
and `navigator` object and what their values are.

In this case, the log in fails because `navigator.plugins` and
`navigator.mimeTypes` now returns the values hard-coded in the HTML
specification instead of being empty arrays.

This is a site issue so I am adding a quirk for now and contacted evangelism.

* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::initializePluginAndMimeTypeArrays):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldNavigatorPluginsBeEmpty const):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/257407@main
